dev/bots/analyze.dart

```diff
@@ -539,65 +500,533 @@ Future<void> verifyInternationalizations() async {
-      ..writeln(materialGenResult.stdout.trim())
-      ..writeln('>>>>>>> gen_localizations')
-      ..writeln('The contents of $materialLocalizationsFile are different from that produced by gen_localizations.')
-      ..writeln()
-      ..writeln('Did you forget to run gen_localizations.dart after updating a .arb file?');
-    exit(1);
+    exitWithError(<String>[
+      '<<<<<<< $materialLocalizationsFile',
+      expectedMaterialResult.trim(),
+      '=======',
+      materialGenResult.stdout.trim(),
+      '>>>>>>> gen_localizations',
+      'The contents of $materialLocalizationsFile are different from that produced by gen_localizations.',
+      '',
+      'Did you forget to run gen_localizations.dart after updating a .arb file?',
+    ]);
   }
   if (cupertinoGenResult.stdout.trim() != expectedCupertinoResult.trim()) {
-    stderr
-      ..writeln('<<<<<<< $cupertinoLocalizationsFile')
-      ..writeln(expectedCupertinoResult.trim())
-      ..writeln('=======')
-      ..writeln(cupertinoGenResult.stdout.trim())
-      ..writeln('>>>>>>> gen_localizations')
-      ..writeln('The contents of $cupertinoLocalizationsFile are different from that produced by gen_localizations.')
-      ..writeln()
-      ..writeln('Did you forget to run gen_localizations.dart after updating a .arb file?');
-    exit(1);
+    exitWithError(<String>[
+      '<<<<<<< $cupertinoLocalizationsFile',
+      expectedCupertinoResult.trim(),
+      '=======',
+      cupertinoGenResult.stdout.trim(),
+      '>>>>>>> gen_localizations',
+      'The contents of $cupertinoLocalizationsFile are different from that produced by gen_localizations.',
+      '',
+      'Did you forget to run gen_localizations.dart after updating a .arb file?',
+    ]);
   }
 }
 
-Future<void> verifyNoTrailingSpaces() async {
-  if (!Platform.isWindows) {
-    final String commitRange = Platform.environment.containsKey('TEST_COMMIT_RANGE')
-        ? Platform.environment['TEST_COMMIT_RANGE']
-        : await _getCommitRange();
-    final List<String> fileTypes = <String>[
-      '*.dart', '*.cxx', '*.cpp', '*.cc', '*.c', '*.C', '*.h', '*.java', '*.mm', '*.m', '*.yml',
-    ];
-    final EvalResult changedFilesResult = await _evalCommand(
-      'git', <String>['diff', '-U0', '--no-color', '--name-only', commitRange, '--', ...fileTypes],
-      workingDirectory: flutterRoot,
+Future<void> verifyNoTrailingSpaces(String workingDirectory, { int minimumMatches = 4000 }) async {
+  final List<File> files = _allFiles(workingDirectory, null, minimumMatches: minimumMatches)
+    .where((File file) => path.basename(file.path) != 'serviceaccount.enc')
+    .where((File file) => path.basename(file.path) != 'Ahem.ttf')
+    .where((File file) => path.extension(file.path) != '.snapshot')
+    .where((File file) => path.extension(file.path) != '.png')
+    .where((File file) => path.extension(file.path) != '.jpg')
+    .where((File file) => path.extension(file.path) != '.jar')
+    .toList();
```

Comment:
Isn't this O^6?  Or is the compiler really able to fold these into a single test on one iteration?

Reply:
I would assume it's O(N), why would it be O(N^6)? Where is just a mapping iterator, it walks the incoming iterator, filters it, and only yields matching lines.

I don't think the compiler knows anything here, it's just how the iterator API is implemented.

Reply:
Sorry, it's not O(N^6), but it's close to O(N*6) (a bit shy of that since each iteration eliminates at least one element) - each of those `where` clauses has to iterate the entire yielded set from the previous one, and those first few are only eliminating one or two files a piece.

You'd save a bunch of iterations if this was just a single `where` with a larger `path.basename(file.path) != ... &&`, or maybe even just `!file.path.endsWith(...) && !file.path.endsWith...`

Reply:
This code runs in milliseconds, I'm not really that concerned about making it faster, certainly not at the cost of readability. :-)